File: ts/src/star-make-ready-v1.ts
import { z } from 'zod';

import { sha256Digest } from './factory/digest.js';

const DigestSchema = z
  .string()
  .regex(/^sha256:[0-9a-f]{64}$/, 'digest must be sha256:<64 lowercase hex>');
const UuidSchema = z
  .string()
  .regex(
    /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/,
    'UUID must be canonical lowercase RFC 4122',
  );
const PositiveRevision = z.number().int().safe().positive();
const OpaqueIdSchema = z
  .string()
  .transform((value) => value.trim())
  .refine(
    (value) => (
      /^[A-Za-z0-9][A-Za-z0-9._/-]{0,254}$/.test(value)
      && value.split('/').every((segment) => (
        segment !== '' && segment !== '.' && segment !== '..'
      ))
    ),
    'technical id must use the opaque id grammar',
  );

const REQUEST_FIELDS = [
  'contract_version',
  'workspace_id',
  'run_id',
  'run_revision',
] as const;
const COMMAND_ID_FIELDS = [
  'workspace_id',
  'run_id',
  'run_revision',
  'request_digest',
  'character_lock_digest',
  'character_lock_version',
  'product_lock_digest',
  'artemis_approval_receipt_id',
  'artemis_approval_receipt_digest',
  'artemis_approval_state_revision',
] as const;
const RECEIPT_FIELDS = [
  'contract_version',
  'command_id',
  ...COMMAND_ID_FIELDS,
  'state',
  'provider_call',
] as const;

function digestFields(
  value: Record<string, unknown>,
  fields: readonly string[],
): string {
  return sha256Digest(
    Object.fromEntries(fields.map((field) => {
      if (!(field in value) || value[field] === undefined) {
        throw new Error(`${field} is required for digest`);
      }
      return [field, value[field]];
    })),
  );
}

function requireOpaqueReceiptId(value: Record<string, unknown>): void {
  const raw = value.artemis_approval_receipt_id;
  const parsed = OpaqueIdSchema.parse(raw);
  if (parsed !== raw) {
    throw new Error('artemis_approval_receipt_id must be canonical');
  }
}

export function deriveStarMakeReadyRequestDigestV1(
  value: Record<string, unknown>,
): string {
  return digestFields(value, REQUEST_FIELDS);
}

export function deriveStarMakeReadyCommandIdV1(
  value: Record<string, unknown>,
): string {
  requireOpaqueReceiptId(value);
  return sha256Digest({
    command_kind: 'star.make_ready',
    ...Object.fromEntries(COMMAND_ID_FIELDS.map((field) => {
      if (!(field in value) || value[field] === undefined) {
        throw new Error(`${field} is required for command id`);
      }
      return [field, value[field]];
    })),
  });
}

export function deriveStarMakeReadyReceiptDigestV1(
  value: Record<string, unknown>,
): string {
  requireOpaqueReceiptId(value);
  return digestFields(value, RECEIPT_FIELDS);
}

const RequestShape = {
  contract_version: z.literal('StarMakeReadyRequest.v1'),
  workspace_id: UuidSchema,
  run_id: UuidSchema,
  run_revision: PositiveRevision,
};

export const StarMakeReadyRequestV1Schema = z
  .object({
    ...RequestShape,
    request_digest: DigestSchema,
  })
  .strict()
  .superRefine((value, ctx) => {
    let expected;
    try {
      expected = deriveStarMakeReadyRequestDigestV1(value);
    } catch {
      return;
    }
    if (value.request_digest !== expected) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'request_digest does not match make-ready request',
        path: ['request_digest'],
      });
    }
  })
  .transform((value) => Object.freeze(value));

const ReceiptShape = {
  contract_version: z.literal('StarMakeReadyReceipt.v1'),
  command_id: DigestSchema,
  workspace_id: UuidSchema,
  run_id: UuidSchema,
  run_revision: PositiveRevision,
  request_digest: DigestSchema,
  character_lock_digest: DigestSchema,
  character_lock_version: PositiveRevision,
  product_lock_digest: DigestSchema,
  artemis_approval_receipt_id: OpaqueIdSchema,
  artemis_approval_receipt_digest: DigestSchema,
  artemis_approval_state_revision: PositiveRevision,
  state: z.literal('succeeded'),
  provider_call: z.literal('none'),
  receipt_digest: DigestSchema,
};

export const StarMakeReadyReceiptV1Schema = z
  .object(ReceiptShape)
  .strict()
  .superRefine((value, ctx) => {
    const issue = (message: string, path: string[]) => {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message, path });
    };
    let expectedRequest;
    let expectedCommand;
    let expectedReceipt;
    try {
      expectedRequest = deriveStarMakeReadyRequestDigestV1({
        contract_version: 'StarMakeReadyRequest.v1',
        workspace_id: value.workspace_id,
        run_id: value.run_id,
        run_revision: value.run_revision,
      });
      expectedCommand = deriveStarMakeReadyCommandIdV1(value);
      expectedReceipt = deriveStarMakeReadyReceiptDigestV1(value);
    } catch {
      return;
    }
    if (value.request_digest !== expectedRequest) {
      issue('request_digest does not match make-ready scope', ['request_digest']);
    }
    if (value.command_id !== expectedCommand) {
      issue('command_id does not match make-ready authority', ['command_id']);
    }
    if (value.receipt_digest !== expectedReceipt) {
      issue('receipt_digest does not match make-ready receipt', ['receipt_digest']);
    }
  })
  .transform((value) => Object.freeze(value));

export type StarMakeReadyRequestV1 = z.infer<
  typeof StarMakeReadyRequestV1Schema
>;
export type StarMakeReadyReceiptV1 = z.infer<
  typeof StarMakeReadyReceiptV1Schema
>;

export type StarMakeReadyAuthorityV1 = Readonly<{
  command_id: string;
  workspace_id: string;
  run_id: string;
  run_revision: number;
  character_lock_digest: string;
  character_lock_version: number;
  product_lock_digest: string;
  artemis_approval_receipt_id: string;
  artemis_approval_receipt_digest: string;
  artemis_approval_state_revision: number;
}>;

export interface StarMakeReadyResolverV1 {
  isCurrentMakeReady(authority: StarMakeReadyAuthorityV1): boolean;
}

export function starMakeReadyReceiptAuthorizesV1(
  receipt: StarMakeReadyReceiptV1,
  resolver: StarMakeReadyResolverV1,
): boolean {
  return resolver.isCurrentMakeReady({
    command_id: receipt.command_id,
    workspace_id: receipt.workspace_id,
    run_id: receipt.run_id,
    run_revision: receipt.run_revision,
    character_lock_digest: receipt.character_lock_digest,
    character_lock_version: receipt.character_lock_version,
    product_lock_digest: receipt.product_lock_digest,
    artemis_approval_receipt_id: receipt.artemis_approval_receipt_id,
    artemis_approval_receipt_digest:
      receipt.artemis_approval_receipt_digest,
    artemis_approval_state_revision:
      receipt.artemis_approval_state_revision,
  }) === true;
}

File: hiob_contracts/star_make_ready_v1.py
"""Deterministic Star receipt proving one run is ready without a provider call."""

from __future__ import annotations

from typing import Annotated, Any, Literal, Mapping, Protocol

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .ares_script_revision_v1 import DigestStr
from .artemis_product_lock_v1 import OpaqueId
from .factory.digest import sha256_digest

_STRICT_FROZEN = ConfigDict(
    extra="forbid",
    frozen=True,
    strict=True,
    revalidate_instances="always",
)
UuidStr = Annotated[
    str,
    Field(
        pattern=(
            r"^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-"
            r"[89ab][0-9a-f]{3}-[0-9a-f]{12}$"
        ),
        strict=True,
    ),
]
PositiveRevision = Annotated[
    int,
    Field(gt=0, le=9_007_199_254_740_991, strict=True),
]

_REQUEST_FIELDS = (
    "contract_version",
    "workspace_id",
    "run_id",
    "run_revision",
)
_COMMAND_ID_FIELDS = (
    "workspace_id",
    "run_id",
    "run_revision",
    "request_digest",
    "character_lock_digest",
    "character_lock_version",
    "product_lock_digest",
    "artemis_approval_receipt_id",
    "artemis_approval_receipt_digest",
    "artemis_approval_state_revision",
)
_RECEIPT_FIELDS = (
    "contract_version",
    "command_id",
    *_COMMAND_ID_FIELDS,
    "state",
    "provider_call",
)


def _json_mapping(value: Mapping[str, Any] | BaseModel) -> dict[str, Any]:
    return (
        value.model_dump(mode="json")
        if isinstance(value, BaseModel)
        else dict(value)
    )


def _digest_fields(
    value: Mapping[str, Any] | BaseModel,
    fields: tuple[str, ...],
) -> str:
    data = _json_mapping(value)
    return sha256_digest({field: data[field] for field in fields})


def derive_star_make_ready_request_digest_v1(
    value: Mapping[str, Any] | BaseModel,
) -> str:
    return _digest_fields(value, _REQUEST_FIELDS)


def derive_star_make_ready_command_id_v1(
    value: Mapping[str, Any] | BaseModel,
) -> str:
    data = _json_mapping(value)
    receipt_id = data["artemis_approval_receipt_id"]
    if (
        not isinstance(receipt_id, str)
        or not receipt_id
        or len(receipt_id) > 255
        or any(
            not (
                character.isascii()
                and (character.isalnum() or character in "._/-")
            )
            for character in receipt_id
        )
        or any(segment in {"", ".", ".."} for segment in receipt_id.split("/"))
    ):
        raise ValueError("artemis_approval_receipt_id is not an opaque id")
    return sha256_digest(
        {
            "command_kind": "star.make_ready",
            **{field: data[field] for field in _COMMAND_ID_FIELDS},
        }
    )


def derive_star_make_ready_receipt_digest_v1(
    value: Mapping[str, Any] | BaseModel,
) -> str:
    derive_star_make_ready_command_id_v1(value)
    return _digest_fields(value, _RECEIPT_FIELDS)


class StarMakeReadyResolverV1(Protocol):
    """Durable authority required to use a structurally valid receipt."""

    def is_current_make_ready(
        self,
        *,
        command_id: str,
        workspace_id: str,
        run_id: str,
        run_revision: int,
        character_lock_digest: str,
        character_lock_version: int,
        product_lock_digest: str,
        artemis_approval_receipt_id: str,
        artemis_approval_receipt_digest: str,
        artemis_approval_state_revision: int,
    ) -> bool: ...


class StarMakeReadyRequestV1(BaseModel):
    """Server-scoped read request; it accepts no authority or provider payload."""

    model_config = _STRICT_FROZEN

    contract_version: Literal["StarMakeReadyRequest.v1"]
    workspace_id: UuidStr
    run_id: UuidStr
    run_revision: PositiveRevision
    request_digest: DigestStr

    @model_validator(mode="after")
    def _validate_digest(self) -> "StarMakeReadyRequestV1":
        if self.request_digest != derive_star_make_ready_request_digest_v1(self):
            raise ValueError("request_digest does not match make-ready request")
        return self


class StarMakeReadyReceiptV1(BaseModel):
    """Read receipt binding current CharacterLock and Artemis product authority."""

    model_config = _STRICT_FROZEN

    contract_version: Literal["StarMakeReadyReceipt.v1"]
    command_id: DigestStr
    workspace_id: UuidStr
    run_id: UuidStr
    run_revision: PositiveRevision
    request_digest: DigestStr
    character_lock_digest: DigestStr
    character_lock_version: PositiveRevision
    product_lock_digest: DigestStr
    artemis_approval_receipt_id: OpaqueId
    artemis_approval_receipt_digest: DigestStr
    artemis_approval_state_revision: PositiveRevision
    state: Literal["succeeded"]
    provider_call: Literal["none"]
    receipt_digest: DigestStr

    @model_validator(mode="after")
    def _validate_bindings(self) -> "StarMakeReadyReceiptV1":
        request_payload = {
            "contract_version": "StarMakeReadyRequest.v1",
            "workspace_id": self.workspace_id,
            "run_id": self.run_id,
            "run_revision": self.run_revision,
        }
        if self.request_digest != derive_star_make_ready_request_digest_v1(
            request_payload
        ):
            raise ValueError("request_digest does not match make-ready scope")
        if self.command_id != derive_star_make_ready_command_id_v1(self):
            raise ValueError("command_id does not match make-ready authority")
        if self.receipt_digest != derive_star_make_ready_receipt_digest_v1(self):
            raise ValueError("receipt_digest does not match make-ready receipt")
        return self

    def authorizes(self, *, resolver: StarMakeReadyResolverV1) -> bool:
        """Require current server state; the receipt itself is not authority."""

        return resolver.is_current_make_ready(
            command_id=self.command_id,
            workspace_id=self.workspace_id,
            run_id=self.run_id,
            run_revision=self.run_revision,
            character_lock_digest=self.character_lock_digest,
            character_lock_version=self.character_lock_version,
            product_lock_digest=self.product_lock_digest,
            artemis_approval_receipt_id=self.artemis_approval_receipt_id,
            artemis_approval_receipt_digest=(
                self.artemis_approval_receipt_digest
            ),
            artemis_approval_state_revision=(
                self.artemis_approval_state_revision
            ),
        ) is True


__all__ = [
    "StarMakeReadyRequestV1",
    "StarMakeReadyReceiptV1",
    "StarMakeReadyResolverV1",
    "derive_star_make_ready_request_digest_v1",
    "derive_star_make_ready_command_id_v1",
    "derive_star_make_ready_receipt_digest_v1",
]
